diff -u xterm-304/button.c xterm/button.c
--- xterm-304/button.c	2016-05-30 15:42:44.000000000 -0400
+++ xterm/button.c	2016-12-16 21:43:10.824319691 -0500
@@ -2408,6 +2408,62 @@
     }
 }
 
+/* ARGSUSED */
+void
+HandleSelectNeedle(Widget w,
+		   XEvent *event GCC_UNUSED,
+		   String *params,
+		   Cardinal *num_params)
+{
+    XtermWidget xw;
+    String needle;
+    size_t len;
+
+    if (!num_params || *num_params < 1) {
+      return;
+    }
+
+    needle = params[0];
+    len = strlen(needle);
+
+    if ((xw = getXtermWidget(w)) != 0) {
+        TScreen *screen = TScreenOf(xw);
+        CELL cell = screen->cursorp;
+        LineData *ld;
+
+        if (!isSameCELL(&screen->startH, &screen->endH)) {
+          /* Something's highlighted, start a row above */
+          cell.row = screen->startH.row - 1;
+          cell.col = screen->max_col;
+        } else {
+          /* Search the whole screen */
+          cell.row = screen->max_row;
+          cell.col = screen->max_col;
+        }
+
+        for (; cell.row >= 0; cell.row--) {
+          ld = getLineData(screen, cell.row);
+
+          for (cell.col = screen->max_col - len; cell.col >= 0; cell.col--) {
+            size_t i;
+
+            for (i = 0; i < len; i++) {
+              if (ld->charData[cell.col + i] != needle[i]) {
+                break;
+              }
+            }
+
+            if (i == len) { /* needle found */
+              screen->selectUnit = Select_WORD;
+              ComputeSelect(xw, &cell, &cell, False);
+              return;
+            }
+          }
+        }
+        DisownSelection(xw);
+    }
+}
+
 static void
 TrackDown(XtermWidget xw, XButtonEvent *event)
 {
diff -u xterm-304/charproc.c xterm/charproc.c
--- xterm-304/charproc.c	2014-05-02 23:48:33.000000000 +0200
+++ xterm/charproc.c	2014-05-14 14:19:33.093237072 +0200
@@ -263,6 +263,7 @@
     { "select-cursor-start",	HandleKeyboardSelectStart },
     { "select-end",		HandleSelectEnd },
     { "select-extend",		HandleSelectExtend },
+    { "select-needle",		HandleSelectNeedle },
     { "select-set",		HandleSelectSet },
     { "select-start",		HandleSelectStart },
     { "send-signal",		HandleSendSignal },
diff -u xterm-304/xterm.h xterm/xterm.h
--- xterm-304/xterm.h	2014-05-03 21:30:17.000000000 +0200
+++ xterm/xterm.h	2014-05-14 14:19:43.483235439 +0200
@@ -804,6 +804,7 @@
 extern void HandleKeyboardStartExtend  PROTO_XT_ACTIONS_ARGS;
 extern void HandleSelectEnd            PROTO_XT_ACTIONS_ARGS;
 extern void HandleSelectExtend         PROTO_XT_ACTIONS_ARGS;
+extern void HandleSelectNeedle         PROTO_XT_ACTIONS_ARGS;
 extern void HandleSelectSet            PROTO_XT_ACTIONS_ARGS;
 extern void HandleSelectStart          PROTO_XT_ACTIONS_ARGS;
 extern void HandleStartExtend          PROTO_XT_ACTIONS_ARGS;
diff -u xterm-304/xterm.man xterm/xterm.man
--- xterm-304/xterm.man	2014-04-14 20:42:54.000000000 +0200
+++ xterm/xterm.man	2014-05-14 15:00:15.749856506 +0200
@@ -5920,6 +5920,10 @@
 This action tracks the pointer and extends the selection.
 It should only be bound to Motion events.
 .TP 8
+.B "select-needle(\fIneedle\fP)"
+This action makes a word selection on the first occurrence of the needle string
+preceding the current selection, if any, or the cursor.
+.TP 8
 .B "select-set()"
 This action stores text that corresponds to the current selection,
 without affecting the selection mode.
